intake/source/discovery.py

```diff
@@ -9,22 +9,253 @@
+                             name, cls.__module__, cls.__name__)
+            continue
+        module_name, object_name = dotted_object_name.rsplit('.', 1)
+        entrypoint = entrypoints.EntryPoint(name, module_name, object_name)
+        logger.debug("Discovered config-specified '%s = %s.%s'",
+                     entrypoint.name,
+                     entrypoint.module_name,
+                     entrypoint.object_name)
+        if name in group:
+            shadowed = group[name]
+            logger.debug("Config shadowed entrypoint '%s = %s.%s'",
+                         shadowed.name,
+                         shadowed.module_name,
+                         shadowed.object_name)
+        if name in package_scan_results:
+            cls = package_scan_results[name]
+            del package_scan_results[name]
+            logger.debug("Config shadowed package scan result '%s = %s.%s'",
+                         name, cls.__module__, cls.__name__)
+        group[name] = entrypoint
+
+    # Discovery is complete.
+
+    if package_scan_results:
+        warnings.warn(
+            f"The drivers {package_scan_results} do not specify entry_points "
+            f"and were only discovered via a package scan. This may break in a "
+            f"future release of intake. The packages should be updated.",
+            FutureWarning)
+
+    # Load entrypoints. Any that were shadowed or banned have already been
+    # removed above.
+    drivers = {}
+    for entrypoint in group.values():
+        try:
+            drivers[entrypoint.name] = _load_entrypoint(entrypoint)
+        except ConfigurationError:
+            logger.exception()
+            continue
+        logger.debug("Loaded entrypoint '%s = %s.%s'",
+                     entrypoint.name,
+                     entrypoint.module_name,
+                     entrypoint.object_name)
 
-    This function searches path (or sys.path) for packages with names that
-    start with plugin_prefix.  Those modules will be imported and scanned for
-    subclasses of intake.source.base.Plugin.  Any subclasses found will be
-    instantiated and returned in a dictionary, with the plugin's name attribute
-    as the key.
+    # Now include any package scan results. Any that were shadowed or
+    # banned have already been removed above.
+    for name, cls in package_scan_results.items():
+        drivers[name] = cls
+        logger.debug("Loaded package scan result '%s = %s.%s'",
+                     name,
+                     cls.__module__,
+                     cls.__name__)
+
+    return drivers
+
+
+def autodiscover_all(path=None, plugin_prefix='intake_', do_package_scan=True):
+    """Discover intake drivers including those registered for the same name.
+
+    Parameters
+    ----------
+    path : str or None
+        Default is ``sys.path``.
+    plugin_prefix : str
+        DEPRECATED. Default is 'intake_'.
+    do_package_scan : boolean
+        Default is True.
```

Comment:
Will this be deprecated as well? It seems like the default will at least be changed at some point. Not sure how best to expose that though.

Reply:
Good catch. Added in dcc84a0. Thanks!